eth/downloader/downloader.go

```diff
@@ -1360,7 +1360,9 @@ func (d *Downloader) processHeaders(origin uint64, td, ttd *big.Int, beaconMode
 						for i, header := range chunkHeaders {
 							td = new(big.Int).Add(td, header.Difficulty)
 							if td.Cmp(ttd) >= 0 {
-								// Terminal total difficulty reached, allow the last header in
+								// Terminal total difficulty reached, write terminal block hash
+								rawdb.WriteTerminalBlockHash(d.stateDB, header.Hash())
+								// Allow the last header in
```

Comment:
This is wrong, at this point I only checked that the diff is > ttd, it might be already a second block on top of ttd. There's no block import dont at this point, just pre-validation. Please don't modify the database. It should be done in `core.Blockchain`, not in the downloader.